components/range_cache_memory_engine/src/range_stats.rs

```diff
@@ -175,98 +192,228 @@ impl RangeStatsManager {
-                    {
-                        let mut mut_prev_top_regions = self.prev_top_regions.lock();
-                        let _ = mut_prev_top_regions.insert(id, region.clone());
+
+        let cached_region_ids = cached_region_ids.into_iter().collect::<HashSet<u64>>();
+        let (mvcc_amplification_to_filter, regions_to_load) = {
+            let mut max_mvcc_amplification: f64 = 0.0;
+            let mut record = self.mvcc_amplification_record.lock();
+            let regions_to_load = curr_top_regions
+                .iter()
+                .filter_map(|(id, (r, region_stats))| {
+                    if !cached_region_ids.contains(id) {
+                        max_mvcc_amplification = max_mvcc_amplification
+                            .max(region_stats.cop_detail.mvcc_amplification());
+                        record.insert(*id, region_stats.cop_detail.mvcc_amplification());
+                        Some(r.clone())
+                    } else {
                         None
                     }
-                    _ => Some(region.clone()),
-                }
+                })
+                .collect();
+            // `max_mvcc_amplification / ma_mvcc_amplification_reduction` is the
+            // expected mvcc amplification factor after cache. Make the half of it to filter
+            // the cached regions.
+            (
+                max_mvcc_amplification / f64::max(1.0, ma_mvcc_amplification_reduction / 2.0),
+                regions_to_load,
+            )
+        };
+
+        info!(
+            "IME mvcc amplification reduction filter";
+            "mvcc_amplification_to_filter" => mvcc_amplification_to_filter,
+        );
+
+        {
+            // TODO(SpadeA): remove it after it's stable
+            let debug: Vec<_> = regions_activity
+                .iter()
+                .map(|(r, s)| {
+                    format!(
+                        "region_id={}, cop={}, cop_detail={:?}, mvcc_amplification={}",
+                        r.id,
+                        s.query_stats.coprocessor,
+                        s.cop_detail,
+                        s.cop_detail.mvcc_amplification(),
+                    )
+                })
+                .collect();
+            info!(
+                "ime collect regions activities";
+                "regions" => ?debug,
+            );
+        }
+
+        let reach_stop_load = memory_controller.reached_stop_load_limit();
+        let mut regions_loaded = self.region_loaded_at.write().unwrap();
+        // Evict at most 1/10 of the regions.
+        let max_count_to_evict = usize::max(1, regions_activity.len() / 10);
+        // Use top MIN_REGION_COUNT_TO_EVICT regions next and prev to filter regions
+        // with very few next and prev. If there's less than or equal to
+        // MIN_REGION_COUNT_TO_EVICT regions, do not evict any one.
+        let regions_to_evict = if regions_activity.len() > MIN_REGION_COUNT_TO_EVICT {
```

Comment:
Not sure if we need to consider the current memory usage(available memory from stop_load_threshold) to determine how many regions need to be evicted and loaded. For example, if the current loaded regions are still hot, but there are incoming regions that both mvcc amp and qps are more the 2 times higher, due to the current load evict logic, seems we can't ensure these new hot regions can be loaded if the memory usage is already near stop-load-threshold.

Reply:
It's very inaccurate to estimate the memory usage of a cached region, at least now. So I does not consider that here.
For hot regions that may not be loaded this time, if it is still hot, it still has chance to be loaded in the next round.